beacon-chain/types/block.go

```diff
@@ -142,34 +141,48 @@ func (b *Block) isSlotValid() bool {
 	return clock.Now().After(validTimeThreshold)
 }
 
-// IsValid is called to decide if an incoming p2p block can be processed.
-// It checks the slot against the system clock, and the validity of the included attestations.
-// Existence of the parent block and the PoW chain block is checked outside of this function because they require additional dependencies.
+// IsValid is called to decide if an incoming p2p block can be processed. It checks for following conditions:
+// 1.) Ensure parent processed.
+// 2.) Ensure pow_chain_ref processed.
+// 3.) Ensure local time is large enough to process this block's slot.
+// 4.) Attestation from proposer of block was included with the block in the network message.
 func (b *Block) IsValid(aState *ActiveState, cState *CrystallizedState, parentSlot uint64) bool {
 	_, err := b.Hash()
 	if err != nil {
-		log.Debugf("Could not hash incoming block: %v", err)
+		log.Errorf("Could not hash incoming block: %v", err)
 		return false
 	}
 
 	if b.SlotNumber() == 0 {
-		log.Debug("Can not process a genesis block")
+		log.Error("Can not process a genesis block")
 		return false
 	}
 
 	if !b.isSlotValid() {
-		log.Debugf("slot of block is too high: %d", b.SlotNumber())
+		log.Errorf("Slot of block is too high: %d", b.SlotNumber())
 		return false
 	}
 
+	// verify proposer from last slot is in one of the AttestationRecord.
+	_, proposerIndex, err := casper.GetProposerIndexAndShard(
+		cState.ShardAndCommitteesForSlots(),
+		cState.LastStateRecalc(),
+		parentSlot)
+	if err != nil {
+		log.Errorf("Can not get proposer index %v", err)
+		return false
+	}
 	for index, attestation := range b.Attestations() {
 		if !b.isAttestationValid(index, aState, cState, parentSlot) {
 			log.Debugf("attestation invalid: %v", attestation)
 			return false
 		}
+		if utils.BitSetCount(attestation.AttesterBitfield) == 1 && utils.CheckBit(attestation.AttesterBitfield, int(proposerIndex)) {
```

Comment:
Start here to review confirmation of proposer attestation

Reply:
This is going to exit early and skip validation of the remaining attesters. This needs to validate all attesters AND confirm that one of them is the parent slot's proposer.